dlrover/python/master/node/worker.py

```diff
@@ -314,103 +315,141 @@ def verify_restarting_training(self, node_id):
-        inextricable pending issue.
-
-        There is 3 main conditions:
+        To prevent training hang by pending workers. Should exit when there
+        is inextricable pending issue.
+
+        The fail strategy:
+        torch:
+        0: skip judgement
+        1 and 2: all workers should be ready within the timeout period
+        tf:
+        0: skip judgement
+        1: at least 1 worker(chief) should be ready within the timeout period
+        2: all role nodes should be ready within the timeout period
+
+        There is 3 main conditions for the judgement:
         1: exist pending nodes
         2: alive nodes number consistently lower than the min nodes requires
         3: 1+2 last for a certain time
 
         Args:
             total_node_num(int): Total node number master managed.
+            job_type(str): Job type. Support AllReduceStrategy and
+                ParameterServerStrategy for now.
 
         Return:
             bool
         """
 
+        # fail strategy
+        strategy = _dlrover_context.pending_fail_strategy
+
         # pending time as timeout for now
         timeout = self._get_pending_timeout()
         logger.debug(
             "Is training hang by pending with total worker "
-            f"num: {total_node_num}, timeout: {timeout}."
+            f"num: {total_node_num}, timeout: {timeout}, strategy: {strategy}."
         )
-        if timeout <= 0:
+        if timeout <= 0 or strategy == 0:
             return False
 
-        cur_nodes = list(self._nodes.values())
-
         # collect pending and running nodes
-        pending_nodes: List[Node] = []
-        running_nodes: List[Node] = []
+        cur_nodes = list(self._nodes.values())
+        pending_workers: List[Node] = []
+        running_workers: List[Node] = []
         for node in cur_nodes:
             if node is None or node.is_released or node.create_time is None:
                 continue
-
             if node.status in [NodeStatus.PENDING, NodeStatus.INITIAL]:
-                pending_nodes.append(node)
+                pending_workers.append(node)
             elif node.status == NodeStatus.RUNNING:
-                running_nodes.append(node)
-        self._pending_nodes = pending_nodes
+                running_workers.append(node)
 
         if not self.has_node_required_info() and total_node_num != len(
-            pending_nodes
+            pending_workers
         ):
             logger.debug(
                 "Skip for no required nodes info and not all nodes pending."
             )
             return False
-        elif 0 < len(pending_nodes) == total_node_num:
-            # all nodes pending
-            logger.info(f"All nodes pending: {pending_nodes}.")
-        else:
-            # partial nodes pending
-            # with condition 1 + 2
+
+        if job_type == DistributionStrategy.ALLREDUCE or (
+            job_type == DistributionStrategy.PS and strategy == 2
```

Comment:
`strategy == 2` replace 2 with some constant like `worker_pending_strategy`. It is easier to understand the meaning of the strategy.

Reply:
lgtm